hc_dpki/src/seed.rs

```diff
@@ -0,0 +1,558 @@
+                Seed {
+                    seed_type: stype.clone(),
+                    seed_buf: buf,
+                }
+            }
+        }
+    }
+
+    /// Generated a mnemonic for the seed.
+    pub fn get_mnemonic(&mut self) -> Result<String, HolochainError> {
+        let entropy = self.seed_buf.read_lock();
+        let e = &*entropy;
+        let mnemonic = Mnemonic::from_entropy(e, Language::English).unwrap();
+        Ok(mnemonic.phrase().to_string())
+    }
+}
+#[derive(Debug)]
+pub struct DevicePinSeed {
+    s: Seed,
+}
+
+impl DevicePinSeed {
+    /// creates bundle for for the seed
+    pub fn get_bundle(
+        &mut self,
+        passphrase: String,
+        hint: String,
+        config: Option<PwHashConfig>,
+    ) -> Result<bundle::KeyBundle, HolochainError> {
+        Ok(self.s.get_seed_bundle(passphrase, hint, config)?)
+    }
+
+    /// delegate to base struct
+    pub fn new(s: SecBuf) -> Self {
+        DevicePinSeed {
+            s: Seed::new(&"hcDevicePinSeed".to_string(), SeedInit(s)),
+        }
+    }
+
+    /// generate an application keypair given an index based on this seed
+    /// @param {number} index
+    /// @return {Keypair}
+    pub fn get_application_keypair(&mut self, index: u64) -> Result<Keypair, HolochainError> {
+        (index >= 1).ok_or(HolochainError::ErrorGeneric("Invalid index".to_string()))?;
+
+        let mut out_seed = SecBuf::with_insecure(32);
+        let mut placeholder = SecBuf::with_insecure_from_string("HCAPPLIC".to_string());
+        kdf::derive(
+            &mut out_seed,
+            index.clone(),
+            &mut placeholder,
+            &mut self.s.seed_buf,
+        )?;
+
+        Ok(Keypair::new_from_seed(&mut out_seed)?)
+    }
+}
+#[derive(Debug)]
+pub struct DeviceSeed {
+    s: Seed,
+}
+
+impl DeviceSeed {
+    /// creates bundle for for the seed
+    pub fn get_bundle(
+        &mut self,
+        passphrase: String,
+        hint: String,
+        config: Option<PwHashConfig>,
+    ) -> Result<bundle::KeyBundle, HolochainError> {
+        Ok(self.s.get_seed_bundle(passphrase, hint, config)?)
+    }
+
+    /// delegate to base struct
+    pub fn new(s: SecBuf) -> Self {
+        DeviceSeed {
+            s: Seed::new(&"hcDeviceSeed".to_string(), SeedInit(s)),
+        }
+    }
+
+    /// generate a device pin seed by applying pwhash of pin with this seed as the salt
+    /// @param {string} pin - should be >= 4 characters 1-9
+    /// @return {DevicePinSeed}
+    pub fn get_device_pin_seed(
+        &mut self,
+        pin: String,
+        config: Option<PwHashConfig>,
+    ) -> Result<DevicePinSeed, HolochainError> {
+        (pin.len() >= 4).ok_or(HolochainError::ErrorGeneric("Invalid PIN Size".to_string()))?;
+
+        // let pin_encoded = base64::encode(&pin);
+        let mut pin_buf = SecBuf::with_insecure_from_string(pin);
+
+        let mut hash = SecBuf::with_insecure(pwhash::HASHBYTES);
```

Comment:
???

Reply:
For now, I guess just change this to `with_secure`... I still like the pattern of taking the outputs as parameters, but in this case, we pretty much always want to protect seeds.

Reply:
Should this be
```suggestion
        let mut hash = SecBuf::with_secure(pwhash::HASHBYTES);
```
since hash is what stores the seed?

Reply:
Ok, done.
I'd suggest later we add `SecureBuf` and `InsecureBuf` to sodium crate, since with what you described it would be a lot more error proof to have the `Seed` struct hold a `SecureBuf` instead of a `SecBuf`. Or maybe have a `SeedBuf` which is secure and hardcoded to 32 bytes ?

Reply:
Its also kind of weird to have an "insecure" SecBuf.